src/Umbraco.Web.Common/Extensions/OAuthOptionsExtensions.cs

```diff
@@ -0,0 +1,80 @@
+    /// that are passed back from the external login provider on failure.
+    /// </summary>
+    public static T SetDefaultErrorEventHandling<T>(this T oAuthOptions, string providerFriendlyName) where T : OAuthOptions
+    {
+        oAuthOptions.Events.OnAccessDenied =
+            context => HandleResponseWithDefaultUmbracoRedirect(context, providerFriendlyName, "OnAccessDenied");
+        oAuthOptions.Events.OnRemoteFailure =
+            context => HandleResponseWithDefaultUmbracoRedirect(context, providerFriendlyName, "OnRemoteFailure");
+
+        return oAuthOptions;
+    }
+
+    private static Task HandleResponseWithDefaultUmbracoRedirect(HandleRequestContext<RemoteAuthenticationOptions> context, string providerFriendlyName, string eventName)
+    {
+        context.SetUmbracoRedirectWithFilteredParams(providerFriendlyName,eventName)
+            .HandleResponse();
+
+        return Task.FromResult(0);
+    }
+
+    /// <summary>
+    /// Sets the context to redirect to the <see cref="SecuritySettings.AuthorizeCallbackErrorPathName"/> path with all parameters, except state, that are passed to the initial server callback configured for the configured external login provider
+    /// </summary>
+    public static T SetUmbracoRedirectWithFilteredParams<T>(this T context, string providerFriendlyName, string eventName)
+        where T : HandleRequestContext<RemoteAuthenticationOptions>
+    {
+        var callbackPath = StaticServiceProvider.Instance.GetRequiredService<IOptions<SecuritySettings>>().Value
+            .AuthorizeCallbackErrorPathName;
+
+        callbackPath = callbackPath.AppendQueryStringToUrl("flow=external-login")
+        .AppendQueryStringToUrl($"provider={providerFriendlyName}")
+        .AppendQueryStringToUrl($"callback-event={eventName}");
+
+        foreach (var oathCallbackErrorParam in oathCallbackErrorParams)
+        {
+            if (context.Request.Query.ContainsKey(oathCallbackErrorParam))
+            {
+                callbackPath = callbackPath.AppendQueryStringToUrl($"{oathCallbackErrorParam}={context.Request.Query[oathCallbackErrorParam]}");
+            }
+        }
+
+        context.Response.Redirect(callbackPath);
+        return context;
+    }
+
+    /// <summary>
+    /// Sets the callbackPath for the RemoteAuthenticationOptions based on the configured Umbraco path and the path supplied.
+    /// By default this will result in "/umbraco/your-supplied-path".
+    /// </summary>
+    /// <param name="options">The options object to set the path on.</param>
+    /// <param name="path">The path that should go after the umbraco path, will add a leading slash if it's missing.</param>
+    /// <returns></returns>
+    public static RemoteAuthenticationOptions SetUmbracoBasedCallbackPath(this RemoteAuthenticationOptions options, string path)
+    {
+        var umbracoCallbackPath = StaticServiceProvider.Instance.GetRequiredService<IOptions<SecuritySettings>>().Value
```

Comment:
🔥  - What is this? It seems unused?, and we should generally not use StaticServiceProvider.Instance for dependency injection.

Reply:
Its used as part of the helpers to make it easier to set the "correct" error event handling when registering externallogin providers.

I would love to know how to do this trough proper DI inside a composer.
example
```csharp
builder.AddBackOfficeExternalLogins(logins =>
        {
            logins.AddBackOfficeLogin(
                backOfficeAuthenticationBuilder =>
                {
                    backOfficeAuthenticationBuilder.AddGitHub(
                        backOfficeAuthenticationBuilder.SchemeForBackOffice(GitHubBackOfficeExternalLoginProviderOptions
                            .SchemeName)!,
                        options =>
                        {
                            options.SetUmbracoBasedCallbackPath("signin-github");
                            options.ClientId = "d261ea94c7bf0255801d";
                            options.ClientSecret = "88e795ccb1050c2806dd9d87ec0c6e2048f1a6a4";
                            options.Scope.Add("user:email");
// this right here                            options.SetDefaultErrorEventHandling(GitHubBackOfficeExternalLoginProviderOptions
                                .SchemeName);
                        });
                });
        });
```

Reply:
The BackOfficeAuthenticationBuilder has the service collection. We should at least provide an overload that do not use the static service provider, and usually we move those using the static service provider into a separate extension file (Friendly*...Extensions) (Even that is it pretty unfriendly to let it be easy to make untestable code).

Thereby we can avoid we use those ourselves.

Reply:
Updated github registration example code 

```csharp
using AspNet.Security.OAuth.GitHub;
using Microsoft.Extensions.Options;
using Umbraco.Cms.Api.Management.Security;
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Web.Common.Helpers;

namespace Umbraco.Cms.Web.UI.Custom;

public class GitHubBackOfficeExternalLoginComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.Services.ConfigureOptions<GitHubBackOfficeExternalLoginProviderOptions>();
        builder.Services.ConfigureOptions<ConfigureGitHubAuthenticationOptions>();

        builder.AddBackOfficeExternalLogins(logins =>
        {
            logins.AddBackOfficeLogin(
                backOfficeAuthenticationBuilder =>
                {
                    backOfficeAuthenticationBuilder.AddGitHub(
                        BackOfficeAuthenticationBuilder.SchemeForBackOffice(GitHubBackOfficeExternalLoginProviderOptions
                            .SchemeName)!,
                        options =>
                        {
                        });
                });
        });
    }
}

public class ConfigureGitHubAuthenticationOptions : IConfigureNamedOptions<GitHubAuthenticationOptions>
{
    private readonly OAuthOptionsHelper _helper;

    public ConfigureGitHubAuthenticationOptions(OAuthOptionsHelper helper)
    {
        _helper = helper;
    }

    public void Configure(GitHubAuthenticationOptions options)
    {
        _helper.SetUmbracoBasedCallbackPath(options, "signin-github");
        options.ClientId = "d261ea94c7bf0255801d";
        options.ClientSecret = "88e795ccb1050c2806dd9d87ec0c6e2048f1a6a4";
        options.Scope.Add("user:email");
        _helper.SetDefaultErrorEventHandling(options, GitHubBackOfficeExternalLoginProviderOptions.SchemeName);
    }

    public void Configure(string? name, GitHubAuthenticationOptions options)
    {
        if (name == BackOfficeAuthenticationBuilder.SchemeForBackOffice(GitHubBackOfficeExternalLoginProviderOptions
                .SchemeName))
        {
            Configure(options);
        }
    }
}

```